Add `RevocationBitmap2022` specification (#879)
iotaledger/identity.rs: documentation/docs/specs/revocation_bitmap_2022.md

---
title: RevocationBitmap2022 Specification
sidebar_label: Revocation Bitmap Specification
description: The specification for the embedded revocation bitmap.
image: /img/Identity_icon.png
keywords:
  - DID
  - specs
  - specifications
  - revocation
  - bitmap
---

# RevocationBitmap2022

## Abstract

> This specification describes an on-tangle mechanism for publishing the revocation status of Verifiable Credentials embedded in an issuer's DID document.

Reviewer: Maybe we should link the first occurrence.
```suggestion
This specification describes an on-tangle mechanism for publishing the revocation status of [Verifiable Credentials](https://www.w3.org/TR/vc-data-model/#dfn-verifiable-credentials) embedded in an issuer's DID document.
```

Author: I'll link our [own VC](https://wiki.iota.org/identity.rs/concepts/verifiable_credentials/overview) section of the wiki, if that's okay.

## Data Model

### Revocation Bitmap Status

For an issuer to enable verifiers to check the status of a verifiable credential, the [`credentialStatus`](https://www.w3.org/TR/vc-data-model/#status) property MUST be specified with the following properties:

- **id**: A [DID URL](https://www.w3.org/TR/did-core/#did-url-syntax) that can be resolved to a service of the issuer which contains the revocation bitmap;
- **type**: the string `"RevocationBitmap2022"`;
- **revocationBitmapIndex**: an unsigned, 32-bit integer expressed as a string, which is the index of the verifiable credential's revocation status in the issuer's revocation bitmap;

### Revocation Bitmap Service

In order to allow verfiers to check the status of a verifiable credential, the DID document of the issuer may contain a service that MUST have the following properties:

- **id**: a URL identifying the revocation bitmap;
- **type**: the string `"RevocationBitmap2022"`;

> - **serviceEndpoint**: a `data` URL [[RFC2397](https://datatracker.ietf.org/doc/html/rfc2397)] embedding a revocation bitmap generated according to the [bitmap generation algorithm](#Bitmap-Generation-Algorithm). The `<mediatype>` MUST be `application/octet-stream` and the `base64` attribute MUST be set.

Author: I guess it's too late now anyway, but I think it's fine?

> - **serviceEndpoint**: a `data` URL [[RFC2397](https://datatracker.ietf.org/doc/html/rfc2397)] embedding a revocation bitmap generated according to the [bitmap generation algorithm](#Bitmap-Generation-Algorithm). The `<mediatype>` MUST be `application/octet-stream` and the `base64` attribute MUST be set.

Reviewer 2: What happens when we place it in a JSON-LD schema like https://www.w3.org/2018/credentials/v1 ?

> - **serviceEndpoint**: a `data` URL [[RFC2397](https://datatracker.ietf.org/doc/html/rfc2397)] embedding a revocation bitmap generated according to the [bitmap generation algorithm](#Bitmap-Generation-Algorithm). The `<mediatype>` MUST be `application/octet-stream` and the `base64` attribute MUST be set.

Author: I honestly don't know how to validate that. I would think the service and `credentialStatus` schema would not overlap, but I'm not familiar enough with JSON-LD to judge properly.

> - **serviceEndpoint**: a `data` URL [[RFC2397](https://datatracker.ietf.org/doc/html/rfc2397)] embedding a revocation bitmap generated according to the [bitmap generation algorithm](#Bitmap-Generation-Algorithm). The `<mediatype>` MUST be `application/octet-stream` and the `base64` attribute MUST be set.

Reviewer 2: I don't either (I did try). I think it's fine for now! We can re-assess if there are problems later.

> ## Algorithms
> 
> The following algorithms describe how to generate and validate revocation bitmaps.

Reviewer 2: ```suggestion
## Algorithms

The following algorithms define how to generate and validate revocation bitmaps.
```

### Bitmap Generation Algorithm

> The following process MUST be followed when producing a `RevocationBitmap2022`:

Reviewer 2: ```suggestion
The following process MUST be followed when producing a `RevocationBitmap2022` to embed in a service endpoint:
```

> 1. Let **bitmap** be a [_roaring bitmap_](https://roaringbitmap.org/) where each bit is initialized to 0 (zero).
> 2. For each revoked credential with an **identifier** not exceeding an unsigned, 32-bit integer, set the bit in **bitmap** at index **identifier** to 1.

Reviewer 2: I would be careful using `identifier` here, the terminology can be confused with `id`. Prefer `index` perhaps?

```suggestion
1. Let **bitmap** be a [_roaring bitmap_](https://roaringbitmap.org/) where each bit is initialized to 0.
2. For each revoked credential with an **index** not exceeding an unsigned, 32-bit integer, set the corresponding bit in **bitmap** at **index** to 1.
```

3. Generate the **bitmap serialization** according to the [roaring bitmap serialization format](https://github.com/RoaringBitmap/RoaringFormatSpec/) using the **bitmap** as input.

> 4. Generate a **compressed bitmap** by using the ZLIB compression algorithm [[RFC 1950](https://datatracker.ietf.org/doc/html/rfc1950)] on the bitmap serialization and base64-encoding [[RFC4648](https://datatracker.ietf.org/doc/html/rfc4648)] the result.

Reviewer 2: ```suggestion
4. Generate a **compressed bitmap** by using the ZLIB compression algorithm [[RFC 1950](https://datatracker.ietf.org/doc/html/rfc1950)] on the **bitmap serialization** and base64-encoding [[RFC4648](https://datatracker.ietf.org/doc/html/rfc4648)] the result.
```

5. Return the **compressed bitmap**.

> 

Reviewer 2: Isn't this missing a step to produce the `Revocation Bitmap Service` itself? I.e. encoding it in the data URL and placing it in the service endpoint? RevocationList2022 has two separate algorithms for the separate steps, but we only include one: https://w3c-ccg.github.io/vc-status-rl-2020/#generate-algorithm

Author: Yes, good point. I renamed those two sections to service endpoint generation/expanstion algorithm, please take look if it's better now.

### Bitmap Expansion Algorithm

> The following process MUST be followed when expanding a compressed revocation list bitmap.
> 
> 1. Let **compressed bitmap** be a compressed revocation bitmap generated using the above algorithm.

Reviewer 2: ```suggestion
The following process MUST be followed when expanding a compressed revocation bitmap.

1. Let **compressed bitmap** be a compressed revocation bitmap generated using the [Bitmap Generation Algorithm](#Bitmap-Generation-Algorithm).
```

> 2. Generate an **uncompressed bitmap** by base64-decoding [[RFC4648](https://datatracker.ietf.org/doc/html/rfc4648)] the **compressed bitmap** and further decompressing using ZLIB [[RFC 1950](https://datatracker.ietf.org/doc/html/rfc1950)].

Reviewer 2: ```suggestion
2. Generate an **uncompressed bitmap** by base64-decoding [[RFC4648](https://datatracker.ietf.org/doc/html/rfc4648)] the **compressed bitmap** and then decompressing the result using ZLIB [[RFC 1950](https://datatracker.ietf.org/doc/html/rfc1950)].
```

3. Generate the **bitmap** by deserializing the **uncompressed bitmap** according to the [roaring bitmap serialization format](https://github.com/RoaringBitmap/RoaringFormatSpec/).
4. Return the **bitmap**.

### Validation Algorithm

The following steps MUST be followed when checking whether a verifiable credential is revoked:

1. Let **credential** be a verifiable credential containing a `credentialStatus` whose `type` is `RevocationBitmap2022`.

> 2. Let **revocation bitmap uri** be the `id` field of the **credential**'s `credentialStatus`.

Reviewer 2: Steps 2 and 3 could probably be switched.

```suggestion
2. Let **revocation bitmap URI** be the `id` field of the **credential**'s `credentialStatus`.
```

> 3. Verify all proofs associated with the **credential**. If a proof fails, return a validation error.

Reviewer 2: There are other properties checked during validation, such as `expirationData` and `issuanceDate`, shouldn't those be mentioned here too (as "other properties of the credential")? I know this was lifted from RevocationList2020, it's just that this kind of overlaps with the general validation specification: https://www.w3.org/TR/vc-data-model/#validation

Author: Indeed, we only have this point because of the corresponding point in `RevocationList2020`, but I find that out of place here, because we can't describe the entire validation procedure anyway. Couldn't we also just remove this and let it be a concern of the validation implementor when they want to check for revocation?

> 4. Resolve the **revocation bitmap uri** to a **revocation bitmap service**, and verify that its type is `RevocationBitmap2022`. Return an error otherwise.

Reviewer 2: ```suggestion
4. Resolve the **revocation bitmap URI** to a **revocation bitmap service** in the issuer's DID Document, and verify that the service `type` is `RevocationBitmap2022`. Return an error otherwise.
```

> 5. Let **compressed bitmap** be the `<data>` part of the `serviceEndpoint` data URL [[RFC2397](https://datatracker.ietf.org/doc/html/rfc2397)] of **revocation bitmap service**.

Reviewer 2: ```suggestion
5. Let **compressed bitmap** be the `data` part of the `serviceEndpoint` data URL [[RFC2397](https://datatracker.ietf.org/doc/html/rfc2397)] in **revocation bitmap service**.
```

  }
}
```

and the corresponding issuer's DID document where credential `"5"` in the `#revocation` service is revoked:

```json
{
  "id": "did:iota:EvaQhPXXsJsGgxSXGhZGMCvTt63KuAFtaGThx6a5nSpw",
  "verificationMethod": [
    {
      "id": "did:iota:EvaQhPXXsJsGgxSXGhZGMCvTt63KuAFtaGThx6a5nSpw#key-1",
      "controller": "did:iota:EvaQhPXXsJsGgxSXGhZGMCvTt63KuAFtaGThx6a5nSpw",
      "type": "Ed25519VerificationKey2018",
      "publicKeyMultibase": "z3hgM9fNkhwgT5mECbj1HdKoFNZgpffwQYEV8WBVHphXq"
    }
  ],
  "capabilityInvocation": [
    {
      "id": "did:iota:EvaQhPXXsJsGgxSXGhZGMCvTt63KuAFtaGThx6a5nSpw#sign-0",
      "controller": "did:iota:EvaQhPXXsJsGgxSXGhZGMCvTt63KuAFtaGThx6a5nSpw",
      "type": "Ed25519VerificationKey2018",
      "publicKeyMultibase": "z83F6zbD3KqaxvQhqo25LvSXzoDdpZmp3EpPVonSVACwZ"
    }
  ],
  "service": [
    {
      "id": "did:iota:EvaQhPXXsJsGgxSXGhZGMCvTt63KuAFtaGThx6a5nSpw#revocation",
      "type": "RevocationBitmap2022",
      "serviceEndpoint": "data:application/octet-stream;base64,ZUp5ek1tQmdZR1NBQUFFZ1ptVUFBQWZPQUlF"
    }
  ]
}
```

## Test Vectors

This section provides test vectors to validate implementations against.

### Test Vector 1

> The following data url decodes to a bitmap of length 0 where no index is revoked:

Reviewer 2: ```suggestion
The following data URL decodes to a bitmap of length 0 where no index is revoked:
```

`"data:application/octet-stream;base64,ZUp5ek1tQUFBd0FES0FCcg=="`

### Test Vector 2

> The following data url decodes to a bitmap of length 3 where indices `5`, `398`, and `67000` are revoked:

Reviewer 2: ```suggestion
The following data URL decodes to a bitmap of length 3 where indices `5`, `398`, and `67000` are revoked:
```

`"data:application/octet-stream;base64,ZUp5ek1tQmdZR0lBQVVZZ1pHQ1FBR0laSUdabDZHUGN3UW9BRXVvQjlB"`.

### Test Vector 3

> The following data url decodes to a bitmap of length 16384 where all indices are revoked:

Reviewer 2: ```suggestion
The following data URL decodes to a bitmap of length 16384 where all indices are revoked:
```

`"data:application/octet-stream;base64,ZUp6dHhERVJBQ0FNQkxESEFWS1lXZkN2Q3E0MmFESmtyMlNrM0ROckFLQ2RBQUFBQUFBQTMzbGhHZm9q"`

## Rationale

> This section describes the rationale behind some of the design decisions of this specification.

Reviewer 2: This could probably be expanded with comparisons to `RevocationList2020` and why some aspects of `RevocationBitmap2022` make it preferable, and the limitations thereof (maximum size).

Author: Expanded that section, please take a look. Would be nice to give precise numbers on how many revocations are possilble with the Tangle message size limits, but I think it's pretty hard to calculate, since (I think) the roaring bitmap's serialization size doesn't scale linearly with the number of contained elements?

Reviewer 2: Its size scales with the number of integer blocks required to represent set bits, conceptually. That's why setting random bits approximates its worst case size, which is true for compression too. I'll take a look later.

Reviewer 2: See https://github.com/iotaledger/identity.rs/pull/879#discussion_r900161631

### Compression

Knowing that messages published to the Tangle cannot exceed [32 KiB](https://github.com/iotaledger/tips/blob/main/tips/TIP-0006/tip-0006.md#message-validation) in size and that the greater the message the more it will cost, the use of compression was assessed.

Although the difference in overall size for the compressed DID document is not significant when adding a compressed bitmap, the size of the decompressed document increases significantly when the bitmap is not compressed. Taking into account that the computational effort to compress and decompress the revocation bitmap would be a cost that only verifiers have to deal with, and that every other user would benefit from a lower memory footprint, we opted for compressing the revocation bitmap.

### Privacy Considerations

Because the revocation bitmap is embedded in the DID document, and thus available without contacting the issuer directly, the issuer cannot correlate how a holder uses their credential. Because of that, it is not necessary for the bitmap to have a minimum size.

### Compressed Bitstring vs. Roaring Bitmap

> Because of their space efficiency, a roaring bitmap is preferred for representing a bitmap in-memory. To avoid the dependency on roaring bitmap, we considered to use a compressed bitstring as the serialization format. However, serialization of such a bitstring is 2-3x slower compared to roaring's serialization format, which becomes an issue on resource-constrained devices (e.g. smartphones) or in web browsers.

Reviewer 2: This should be above the `Privacy Considerations` section.

## Revocation Bitmap Concept

The revocation status of a verifiable credential is expressed as a binary value. The issuer keeps a bitmap of indices corresponding to verifiable credentials they have issued. If the binary value of the index in the bitmap is 1 (one), the verifiable credential is revoked, if it is 0 (zero) it is not revoked.

## Data Model

### Revocation Bitmap Status

For an issuer to enable verifiers to check the status of a verifiable credential, the [`credentialStatus`](https://www.w3.org/TR/vc-data-model/#status) property MUST be specified with the following properties:

| Property | Description |
| :--- | :--- |

> | `id` | The constraints on the `id` property are listed in the [Verifiable Credentials Data Model specification](https://www.w3.org/TR/vc-data-model/). The `id` MUST be a [DID URL](https://www.w3.org/TR/did-core/#did-url-syntax) that resolves to a [Revocation Bitmap Service](#revocation-bitmap-service) in the DID Document of the issuer. |

Author: I've asked for clarification on this [here](https://github.com/w3c-ccg/vc-status-list-2021/issues/31). Since we've released 0.6 already, I think we can merge the spec as-is and revise later if needed.

5. Let **revocation index** be the integer value of the `revocationBitmapIndex` property of the **credential**.
6. Let **revoked** be the value of the bit at index **revocation index** in the **revocation bitmap**.
7. Return `true` if **revoked** is 1, `false` otherwise.

## Test Vectors

This section provides test vectors to validate implementations against.

### Test Vector 1

The following data URL decodes to a bitmap of length 0 where no index is revoked:

`"data:application/octet-stream;base64,ZUp5ek1tQUFBd0FES0FCcg=="`

### Test Vector 2

The following data URL decodes to a bitmap of length 3 where indices `5`, `398`, and `67000` are revoked:

`"data:application/octet-stream;base64,ZUp5ek1tQmdZR0lBQVVZZ1pHQ1FBR0laSUdabDZHUGN3UW9BRXVvQjlB"`.

### Test Vector 3

The following data URL decodes to a bitmap of length 16384 where all indices are revoked:

`"data:application/octet-stream;base64,ZUp6dHhERVJBQ0FNQkxESEFWS1lXZkN2Q3E0MmFESmtyMlNrM0ROckFLQ2RBQUFBQUFBQTMzbGhHZm9q"`

## Rationale

This section describes the rationale behind some of the design decisions of this specification.

### Compression

> Considering that messages published to the Tangle cannot exceed [32 KiB](https://github.com/iotaledger/tips/blob/main/tips/TIP-0006/tip-0006.md#message-validation) in size, and that larger messages have increased requirements, the use of compression was assessed.
> 

Reviewer 2: Maybe we can mention a ballpark figure of how many revocations are possible to fit on the Tangle with compression here?

```suggestion
Considering that messages published to the Tangle cannot exceed [32 KiB](https://github.com/iotaledger/tips/blob/main/tips/TIP-0006/tip-0006.md#message-validation) in size, and that larger messages have increased requirements, the use of compression was assessed.

The precise size of a serialized bitmap varies based on the number and distribution of revoked indices. With compression, it is possible to achieve on the order of 100,000 - 200,000 in a 32KiB DID Document when indices are revoked uniformly randomly, and significantly more if consecutive indices are revoked.
```

### Comparison to `RevocationList2020`

The [RevocationList2020 specification](https://w3c-ccg.github.io/vc-status-rl-2020/) describes a revocation mechanism using a verifiable credential that contains a bitmap, similar to the `RevocationBitmap2022` approach. The credential is hosted outside of the DID document and the verifier thus needs to fetch it from an external resource, likely one controlled by the issuer. This has privacy implications as the issuer can track where a fetch request for the credential came from and infering who the credential was verified by and for what purpose. The issuer can also potentially infer which credential was checked. Because `RevocationBitmap2022` is embedded in the DID document, which can be obtained without the issuer's knowledge, this approach does not suffer from those shortcomings. See also the [privacy considerations](#privacy-considerations).

> 

Reviewer 2: Maybe also mention the minimum size of `RevocationList2020`.

```suggestion

Another difference is that [RevocationList2020](https://w3c-ccg.github.io/vc-status-rl-2020/) specifies a minimum initial size of 131,072 for its bitstring, to mitigate the potential for correlating individuals when few credentials have been issued. `RevocationBitmap2022` uses 
a roaring bitmap instead of a bitstring, meaning there is no minimum size and the maximum index is not fixed (apart from the upper bound of an unsigned 32-bit integer). This means the bitmap cannot be used to correlate small populations without more information not present in the bitmap itself. However, both schemes still reveal publicly how many credentials have been revoked, which could be used to infer other information if more knowledge about how an issuer assigns credential revocation indexes is known.

```

A downside of embedding the revocation list in the DID document is that storage in a distributed ledger (DLT) is usually more expensive than storage outside of it. The DLT might also impose message size limitations, capping the total number of revocations that can be done (see also [compression](#compression)).

### Privacy Considerations

> Because the revocation bitmap is embedded in the DID document, and thus available without contacting the issuer directly, the issuer cannot correlate how a holder uses their credential. Because of that, it is not necessary for the bitmap to have a minimum size.

Reviewer 2: Remove reference to minimum size, see previous comment which suggests adding a separate paragraph for it.

```suggestion
Because the revocation bitmap is embedded in the DID document, and thus available without contacting the issuer directly, the issuer cannot correlate how a holder uses their credential.
```
